tests/conftest.py

```diff
@@ -55,15 +59,119 @@ def start_postgresql_container(self):
             environment=self.db_environment,
             ports=self.db_ports)
 
-    def stop_postgresql_container(self):
+    def stop_container(self):
         """Stop PostgreSQL Container."""
         if self.container is None:
             self.container = self.docker_client.containers.get(
                 self.config.CONTAINER_NAME)
-        self.container.stop()
+
+        if self.container is not None:
+            self.container.stop()
+
+
+def delete_migration_artifacts():
+    print('Deleting migration artifacts...')
+    rootdir = os.path.abspath('./migrations/versions')
+
+    for file in os.listdir(os.fsencode(rootdir)):
+        filename = os.fsdecode(file)
+        if filename.endswith(".py"):
+            os.remove(os.path.join(rootdir, filename))
+        else:
+            continue
+
+
+@pytest.fixture(scope='module')
+def regular_client():
+    """Setup the PostgreSQL database instance and run migrations.
+
+    Returns:
+        client (object): The Flask test client for the application.
+
+    """
+
+    delete_migration_artifacts()
+
+    data_resource_manager = DataResourceManagerSync()
+    app = data_resource_manager.create_app()
+    postgres = PostgreSQLContainer()
+    postgres.start_container()
+
+    data_model_manager = DataModelManagerSync()
+
+    upgraded = False
+    counter = 1
+    counter_max = 10
+    while not upgraded and counter <= counter_max:
+        try:
+            with app.app_context():
+                print("------------- running upgrade")
+                data_model_manager.run_upgrade()
+                print("------------- running monitor data resources")
+                data_resource_manager.monitor_data_resources()
+                print("------------- running monitor data models")
+                data_model_manager.monitor_data_models()
+                
+                data_model_manager.run_upgrade()
+                upgraded = True
+        except Exception as e:
+            print(f"Not upgraded, sleeping... {counter}/{counter_max} time(s)")
+            counter += 1
+            sleep(1)
+
+    if counter > counter_max:
+        print("Max fail reached; stopping postgres container")
+        postgres.stop_container()
+    else:
+        yield app.test_client()
+        postgres.stop_container()
 
 
 @pytest.fixture(scope='module')
-def psql_docker():
-    """Database container."""
-    return PostgreSQLContainer()
+def custom_client():
```

Comment:
I'd like to see this refactored. The `regular` and `custom` client fixtures share soooooooooo much code. Can you consolidate? I can imagine a couple ways. One, you could do a classed-based approach (with `client` as the parent class, with has a property of custom=true). Two, you could have a function that accepts an arg, e.g., custom=true.

I know it's not essential, but a future developer would have an easier time making sense of these fixtures.